docs: Add multichain Safe deployments guide (#617)
safe-global/safe-docs: pages/sdk/protocol-kit/guides/safe-deployment/address-replication.mdx

import { Steps } from "nextra/components";

# Safe Address Replication Across Chains

In this guide, you will learn how to replicate a Safe address across different chains using the Protocol Kit. This process includes initializing the Protocol Kit, configuring the Safe, predicting the Safe address on different chains, and executing the deployment.

We will demonstrate how to deploy a Safe account on both the Sepolia and Chiado chains with the same address using the Protocol Kit.

For more detailed information, see the [Protocol Kit Reference](../../../protocol-kit/reference/safe.mdx).

## Prerequisites

- [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)

## Install dependencies

First, you need to install the Protocol Kit.

```bash

> pnpm add @safe-global/protocol-kit

Reviewer: pnpm instead yarn or npm?

Author: I've updated the installation snippet to align with our new documentation format:

```
<CH.Section>
  <CH.Code style={{ boxShadow: 'none' }}>
    ```bash pnpm
    pnpm add @safe-global/protocol-kit
    ```

    ```bash npm
    npm install @safe-global/protocol-kit
    ```

    ```bash yarn
    yarn add @safe-global/protocol-kit
    ```
  </CH.Code>
</CH.Section>
```

```

## Steps

<Steps>

### Imports

> Begin by importing the `Safe` class from the Protocol Kit and chain configuration from viem.

Reviewer: Same as the comment in deploy-safe

> Begin by importing the `Safe` class from the Protocol Kit and chain configuration from viem.

Reviewer 2: ```suggestion
Begin by importing the `Safe` class from the Protocol Kit and chain configuration from Viem.
```

Author: done!

{/* <!-- vale off --> */}

```typescript
import Safe, { SafeDeploymentConfig, SafeAccountConfig } from '@safe-global/protocol-kit'
import { waitForTransactionReceipt } from 'viem/actions'
import { sepolia, gnosisChiado } from 'viem/chains'
```

{/* <!-- vale on --> */}

### Create a signer

> Firstly, you need to get a signer to execute the deployment transaction in each chain. this example uses a private key, but any way to get an `EIP-1193` compatible signer can be used.

Reviewer: Same as the comment in deploy-safe

{/* <!-- vale off --> */}

```typescript
const SIGNER_PRIVATE_KEY = // ...
```

{/* <!-- vale on --> */}

### Define the Safe Configuration

{/* <!-- vale off --> */}

> Define the configuration for your Safe. This includes the `owners`, `threshold`, deployment type, salt nonce for address predictability, and the version of the Safe.

Reviewer: Link to the protocol kit reference to get more info about the options

Reviewer 2: Some properties are also mentioned as `code` and some others with words. Either one way or the order.

```typescript
// Safe account config
const safeAccountConfig: SafeAccountConfig = {
  owners: ['0xOwnerAddress'],
  threshold: 1
}

// Safe deployment config
const safeDeploymentConfig: SafeDeploymentConfig = {
  deploymentType: 'canonical',
  saltNonce: '12345',
  safeVersion: '1.3.0'
}
```

{/* <!-- vale on --> */}

> ### Initialize the `ProtocolKit` for Sepolia and Chiado chain

Reviewer: ```suggestion
### Initialize the `protocol-kit` for Sepolia and Chiado chain
```

Reviewer 2: ```suggestion
### Initialize the Protocol Kit for Sepolia and Chiado chain
```

When writing about the Protocol Kit we are using the form of full name. The kebap case is only used for imports and code snippets.

{/* <!-- vale off --> */}

Initialize the Protocol Kit for deploying on Sepolia and Chiado. Set up the provider and signer along with the predicted Safe configuration.

> You can use the `getAddress` method to ensure that the Safe account has the same address across chains.

Reviewer: ```suggestion
You can use the `getAddress` method to ensure that the Safe account has the same predicted address across chains.
```

  console.log('safeAddressInSepolia: ', safeAddressInSepolia)
  console.log('safeAddressInChiado: ', safeAddressInChiado)
```

{/* <!-- vale on --> */}

### Deploy the Safe on Sepolia

Create and send the deployment transaction on Sepolia.

{/* <!-- vale off --> */}

```typescript
  const deploymentTransactionInSepolia = await protocolKitSepolia.createSafeDeploymentTransaction()

  const sepoliaClient = await protocolKitSepolia.getSafeProvider().getExternalSigner()

  const txHashSepolia = await sepoliaClient!.sendTransaction({
    to: deploymentTransactionInSepolia.to,
    value: BigInt(deploymentTransactionInSepolia.value),
    data: deploymentTransactionInSepolia.data as `0x${string}`,
    chain: sepolia
  })

  await waitForTransactionReceipt(sepoliaClient!, { hash: txHashSepolia })

  protocolKitSepolia = await protocolKitSepolia.connect({ safeAddress: safeAddressInSepolia })

  // Confirm the Safe is deployed in Sepolia and fetch Safe info
  console.log('Is Safe deployed:', await protocolKitSepolia.isSafeDeployed())
  console.log('Safe Address:', await protocolKitSepolia.getAddress())
```

{/* <!-- vale on --> */}

### Repeat Deployment for Chiado

> Repeat the initialization and deployment for the Chiado chain using the same predictedSafe configuration to ensure the Safe address matches the one deployed on Sepolia.

Reviewer 2: ```suggestion
Repeat the initialization and deployment for Chiado chain using the same `predictedSafe` configuration to ensure the Safe address matches the one deployed on Sepolia.
```
